ctk/common/src/main/kotlin/org/codice/compliance/verification/core/SubjectComparisonVerifier.kt

```diff
@@ -0,0 +1,241 @@
+     */
+    private fun verifyIdContentsMatchAuthnRequest(reqId: Node, resId: Node) {
+        val reqFormat = reqId.getFilteredFormatValue()
+        val resFormat = resId.getFilteredFormatValue()
+
+        // If they share a format value that isn't "unspecified", but have different contents
+        if (reqFormat == resFormat
+                && reqFormat != null
+                && reqId.textContent != resId.textContent) {
+            throw SAMLComplianceException.create(SAMLCore_3_3_4_b2,
+                    message = "One of the Response's Subject identifier's content " +
+                            "[${resId.textContent}] is not identical to the AuthnRequest's " +
+                            "Subject identifier's content [${reqId.textContent}], even though " +
+                            "they share the same identifier format [$reqFormat].",
+                    node = resId)
+        }
+    }
+
+    /**
+     * Verifies that the request SubjectConfirmations **strongly match** the response confirmations.
+     *
+     * Note that this matching only compares the methods of the SubjectConfirmations. In order to
+     * properly strongly match we would need to "confirm" the SubjectConfirmation (Hard To Test)
+     */
+    private fun verifyConfirmationsMatchAuthnRequest(requestConfirmations: List<Node>,
+                                                     responseConfirmations: List<Node>) {
+
+        val requestMethods = requestConfirmations.map { it.attributeText("Method") }.toSet()
+        val responseMethods = responseConfirmations.map { it.attributeText("Method") }.toSet()
+
+        // If none the Request SubjectConfirmations have any matching Response SubjectConfirmations
+        if (requestMethods.none { reqMethod ->
+                    responseMethods.any { resMethod -> reqMethod == resMethod }
+                }
+        ) throw SAMLComplianceException.create(SAMLCore_3_3_4_c2,
+                message = "One of the Response's Subjects had no SubjectConfirmation Methods " +
+                        "[$responseMethods] matching the AuthnRequest's SubjectConfirmation " +
+                        "Methods [$requestMethods].",
+                node = response)
+    }
+
+    /**
+     * Returns the Format attribute value if it isn't empty or "unspecified". Else returns null.
+     *
+     * @param node The {@code Node} attribute to check the format value on
+     * @return The filtered Format attribute value
+     */
+    private fun Node.getFilteredFormatValue(): String? {
+        return this.attributeText(FORMAT)?.let {
+            when (it) {
+                EMPTY_STRING -> null
+                UNSPECIFIED_URI -> null
+                else -> it
+            }
+        }
+    }
+
+    private fun Node.getId(): Node? {
+        return this.children().firstOrNull { it.localName != "SubjectConfirmation" }
+    }
+
+    private fun Node.getSubjectConfirmations(): List<Node>? {
```

Comment:
✏️ I recognize that you're using the `null` state of this value as a quick conditional check, but it's a bit more fragile than returning an empty list. As it's private, it's not a big deal; however, we should try to reduce the nulls we pass around, not increase them.

Reply:
I personally find more value in Kotlin passing around nulls. Nulls offer more alternatives than empty lists. Purely my opinion though.

Reply:
After discussion with Richard this is the conclusion we came to (summarized by @bakejeyner):
`Avoid null values when you can. Kotlin _does_ have fancy null stuff, but only because nulls are poorly handled in Java. Rather than using that syntax to make cool looking code, try not to use it because it is only there to fix the mistakes other languages make.`